api/providers/tasks.py

```diff
@@ -0,0 +1,539 @@
+            draft_id=draft_id, error=repr(e),
+        )
+
+    # Set subjects
+    # TODO: if available, capture specific exceptions during setting subject
+    draft.set_subjects_from_relationships(subject_ids, auth)
+
+    # Set affiliated institutions
+    for institution in affiliated_institutions:
+        try:
+            draft.add_affiliated_institution(institution, initiator)
+        except UserNotAffiliatedError:
+            error = 'Initiator [{}] is not affiliated with institution [{}]'.format(initiator._id, institution._id)
+            raise RegistrationBulkCreationRowError(row.upload.id, row.id, row_title, row_external_id, error=error)
+
+    # Set registration responses
+    # TODO: if available, capture specific exceptions during setting responses
+    draft.update_registration_responses(responses)
+
+    # Set tags
+    # TODO: if available, capture specific exceptions during setting tags
+    draft.update_tags(tags, auth=auth)
+
+    # Set contributors
+    for contributor in contributor_list:
+        email = contributor.get('email')
+        full_name = contributor.get('full_name')
+        if not email or not full_name:
+            error = 'Invalid contributor format: missing email and/or full name'
+            raise RegistrationBulkCreationRowError(row.upload.id, row.id, row_title, row_external_id, error=error)
+        bibliographic = email in author_set
+        permission = ADMIN if email in admin_set else (WRITE if email in read_write_set else READ)
+        try:
+            draft.add_contributor_registered_or_not(
+                auth, full_name=full_name, email=email,
+                permissions=permission, bibliographic=bibliographic,
+            )
+        except ValidationValueError as e:
+            # `add_contributor_registered_or_not` throws several ValidationError / ValidationValueError that
+            # needs to be treated differently.
+            if e.message.endswith(' is already a contributor.'):
+                logger.warning('Contributor already exists: [{}]'.format(email))
+            else:
+                error = 'This contributor cannot be added: [email="{}", error="{}"]'.format(email, e.message)
+                raise RegistrationBulkCreationRowError(row.upload.id, row.id, row_title, row_external_id, error=error)
+        except UserStateError as e:
+            error = 'This contributor cannot be added: [email="{}", error="{}"]'.format(email, repr(e))
+            raise RegistrationBulkCreationRowError(row.upload.id, row.id, row_title, row_external_id, error=error)
+    draft.save()
+    row.is_completed = True
+    row.save()
+    logger.info('Draft registration created: [{}]'.format(row.draft_registration.id))
+
+    # Once draft registration has been created, bulk creation of this row is considered completed.
+    # Any error that happens during auto approval doesn't affect the state of upload job and registration row.
+    if auto_approval:
```

Comment:
so we leave the row as a draft it `auto_approval` isn't on? I would have expected that we would still register the draft, but allow users to approve/reject normally (i.e. skip the `registration.sanction.accept()`). I'm sure that Mark or whomever has weighed in, but that's confusing to me.

Reply:
Oh, I see. Let me check with Mark and I think I must misunderstood the state for "non-auto-approval" registrations.